plugins/language_model/knn.cpp

```diff
@@ -1106,7 +1126,12 @@ namespace {
     }
 
     auto candidates = *maybe_candidates;
-    k = std::min(k, static_cast<uint32_t>(candidates.size()));
+    if (arg_k > 0) {
+      k = std::min(k, static_cast<uint32_t>(candidates.size()));
+    } else {
+      k = static_cast<uint32_t>(
+        std::max(0, static_cast<int32_t>(candidates.size() + arg_k + 1)));
+    }
```

Comment:
Hmm. Should we use `arg_k` here too...?

Reply:
Thanks for your comment. It really helped me organize things.